File: exercises/piece-of-cake/solution-b.py
# Code tested on Python 2.7 in http://repl.it/languages/python-turtle

import turtle

t = turtle.Turtle()

# Jump to a position without drawing
def jumpToPos(pos):
  (x,y) = pos
  t.penup()
  t.setx(x)
  t.sety(y)
  t.pendown()

# Hop forward without drawing
def hop(x):
  t.penup()
  t.forward(x)
  t.pendown()

# Draw a filled rectangle
def drawRect(width, height):
  t.begin_fill()
  t.forward(width)
  t.right(90)
  t.forward(height)
  t.right(90)
  t.forward(width)
  t.right(90)
  t.forward(height)
  t.right(90)
  t.end_fill()

# Slice a rectangle in a number of pieces
def sliceRect(width, height, pieces):
  pieceWidth = float(width)/pieces
  for i in range(pieces-1):
    hop(pieceWidth)
    t.right(90)
    t.forward(height)
    hop(-height)
    t.left(90)
  hop(-pieceWidth*(pieces-1))

# Slice each piece into a number of slices
def slicePieces(width, height, pieces, slices):
  pieceWidth = float(width)/pieces
  for p in range(pieces):
    sliceRect(pieceWidth, height, slices)
    hop(pieceWidth)
  hop(-pieceWidth*(pieces-1))



# Illustrate addition of fractions
def showFractionAdd(n, m):
  # set constants
  t.speed(0)
  t.setheading(0)
  cakeWidth = 300
  cakeHeight = 50
  cake1pos = (-150, 100)
  cake2pos = (-150, 25)
  writepos = (-150, -50)

  # Draw the two cakes
  t.width(3)
  t.color('brown')
  t.fillcolor('pink')
  t.setheading(0)
  jumpToPos(cake1pos)
  drawRect(cakeWidth, cakeHeight)
  jumpToPos(cake2pos)
  drawRect(cakeWidth, cakeHeight)

  # Cut cakes in n and m pieces
  t.color('red')
  t.width(5)
  jumpToPos(cake1pos)
  sliceRect(cakeWidth, cakeHeight, n)
  jumpToPos(cake2pos)
  sliceRect(cakeWidth, cakeHeight, m)

  # Eat one piece from each cake
  t.color('brown')
  t.fillcolor('brown')
  t.width(1)
  jumpToPos(cake1pos)
  drawRect(cakeWidth/n, cakeHeight)
  jumpToPos(cake2pos)
  drawRect(cakeWidth/m, cakeHeight)

  # Slice each piece in m and n slices
  t.color('blue')
  t.width(1)
  jumpToPos(cake1pos)
  slicePieces(cakeWidth, cakeHeight, n, m)
  jumpToPos(cake2pos)
  slicePieces(cakeWidth, cakeHeight, m, n)

  # Write equation
  jumpToPos(writepos)
  s1 = "1/"+str(n)
  s2 = "1/"+str(m)
  s3 = str(n+m)+"/"+str(n*m)
  s4 = s1 + " + " + s2 + " = " + s3
  t.write(s4, font=("Arial", 12, "normal"))

# Example call
showFractionAdd(2,5)

File: exercises/piece-of-cake/solution-b.scala
// Helper function for coding repetition
def repeat(n: Int)(f: => Unit) { 
  if (n > 0) {
    f
    repeat(n-1)(f)
  }
}

// Jump to position with x and y values
def jumpTo(x: Int, y: Int) = {
  setPosition(x,y);
}

// Jump to position with (x,y) value
def jumpToPos(pos:(Int, Int)) = {
  val (x,y) = pos
  setPosition(x,y)
}

// Draw a rectangle
def drawRect(width:Int, height:Int) = {
  forward(width)
  right(90)
  forward(height)
  right(90)
  forward(width)
  right(90)
  forward(height)
  right(90)
}

// Slice a rectangle in a number of pieces
def sliceRect(width:Double, height:Double, pieces:Int) = {
  val pieceWidth = width/pieces
  repeat (pieces-1) {
    hop(pieceWidth)
    right(90)
    forward(height)
    hop(-height)
    left(90)
  }
  hop(-(pieceWidth)*(pieces-1)) // hop back to initial position
}

// Slice each piece into a number of slices
def slicePieces(width:Double, height:Double, pieces:Int, slices:Int) = {
  val pieceWidth = width/pieces
  repeat (pieces) {
    sliceRect(pieceWidth, height, slices)
    hop(pieceWidth)
  }
  hop(-(pieceWidth)*(pieces))
}

def showFractionAdd(n:Int, m:Int) = {
  // set constants
  setAnimationDelay(0)
  setHeading(0)
  val cakeWidth = 300
  val cakeHeight = 50
  val cake1pos = (-50, 75)
  val cake2pos = (-50, 0)
  val writepos = (-50, -75)
  
  // Draw the two cakes
  setPenThickness(3)
  setPenColor(brown)
  setFillColor(pink)
  jumpToPos(cake1pos)
  drawRect(cakeWidth,cakeHeight)
  jumpToPos(cake2pos)
  drawRect(cakeWidth, cakeHeight)

  // Cut cakes in n and m pieces
  setPenColor(red)
  setPenThickness(5)
  jumpToPos(cake1pos)
  sliceRect(cakeWidth, cakeHeight, n)
  jumpToPos(cake2pos)
  sliceRect(cakeWidth, cakeHeight, m)
  
  // Eat one piece from each cake
  setPenColor(brown)
  setFillColor(brown)
  setPenThickness(1)
  jumpToPos(cake1pos)
  drawRect(cakeWidth/n, cakeHeight)
  jumpToPos(cake2pos)
  drawRect(cakeWidth/m, cakeHeight)
  
  // Slice each piece in m and n slices
  setPenColor(blue)
  setPenThickness(1)
  jumpToPos(cake1pos)
  slicePieces(cakeWidth, cakeHeight, n, m)
  jumpToPos(cake2pos)
  slicePieces(cakeWidth, cakeHeight, m, n)

  // Write equation
  jumpToPos(writepos)
  setHeading(90)
  val s1 = "1/"+n.toString
  val s2 = "1/"+m.toString
  val s3 = (n+m).toString + "/" + (n*m).toString
  val s4 = s1 + " + " + s2 + " = " + s3
  write(s4)

}

// Example call
showFractionAdd(2,5)